proxy/src/bind.rs

```diff
@@ -40,19 +40,31 @@ pub struct BindProtocol<C, B> {
     protocol: Protocol,
 }
 
-/// A type of service binding
+/// A bound service that can re-bind itself on demand.
+///
+/// Reasons this would need to re-bind:
+///
+/// - `BindsPerRequest` can only service 1 request, and then needs to bind a
+///   new service.
+/// - If there is an error in the inner service (such as a connect error), we
+///   need to throw it away and bind a new service.
+pub struct BoundService<B: tower_h2::Body + 'static> {
+    bind: Bind<Arc<ctx::Proxy>, B>,
```

Comment:
seems to me that, since `Bind::new_bound_service` takes `&self`, this could be
```rust
pub struct BoundService<'a, B: tower_h2::Body + 'static> {
    bind: &'a Bind<Arc<ctx::Proxy>, B>,
    // ...
}
```

and then we wouldn't have to clone in `new_bound_service`?